### PR TITLE
shapefile reader for polygons

### PR DESCRIPTION
**Pull request by zhangjianting**
1. Please write a description in this text box of the changes that are being
   made.
In addition to existing readers that read files in SoA format directly, a shapefile reader for polygons is added which takes a shapefile name and returns four gdf columns for polygons. 
2. Please ensure that you have written units tests for the changes made/features
   added.
Yes.

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -293,7 +296,7 @@ endif(USE_NVTX)
 ###################################################################################################
 # - link libraries --------------------------------------------------------------------------------
 
-target_link_libraries(cuspatial cudf rmm cudart cuda nvrtc)
+target_link_libraries(cuspatial cudf rmm cudart cuda nvrtc gdal)
```
Comment (harrism) [inline cpp/CMakeLists.txt]: Please use the cmake module FindGDAL to ensure goal is found correctly.
https://cmake.org/cmake/help/latest/module/FindGDAL.html

File: cpp/include/cuspatial/shp_readers.hpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * @param[out] ply_fpos: index polygons: prefix sum of number of rings of all
+ *             polygons
+ * @param[out] ply_rpos: index rings: prefix sum of number of vertices of all
+ *             rings
+ * @param[out] ply_x: x coordinates of concatenated polygons
+ * @param[out] ply_y: y coordinates of concatenated polygons
+ *
+ * @note: x/y can be lon/lat.
+**/
+void read_polygon_shp(const char *filename,
```
Comment (harrism) [inline cpp/include/cuspatial/shp_readers.hpp]:
```suggestion
void read_polygon_shapefile(const char *filename,
```

File: cpp/include/cuspatial/shp_readers.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+#pragma once
+
+typedef struct gdf_column_ gdf_column; // forward declaration
+
+namespace cuspatial {
+
+/**
+ * @brief read polygon data from file in SoA format
```
Comment (harrism) [inline cpp/include/cuspatial/shp_readers.hpp]: I don't think SoA format is a standard terminology in GIS, is it? If not, let's not use jargon if we can avoid it. 

```suggestion
 * @brief read polygon data from an ESRI Shapefile.
```

File: cpp/src/io/shp/polygon_shp_reader.cu [hunk 4]
```diff
@@ -0,0 +1,243 @@
+/*
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shp_reader.cu]: Filename: polygon_shapefile_reader.cu

File: cpp/src/io/shp/polygon_shp_reader.cu [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+#include <cuspatial/shp_readers.hpp>
+#include <utility/utility.hpp>
+
+#include <ogrsf_frmts.h>
+
+using namespace std;
+
+namespace cuspatial
+{
+
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shp_reader.cu]: Only public APIs should be in namespace `cuspatial`. Functions only used inside this file should be in an anonymous namespace. Functions used within cuspatial across files should be in `cuspatial::detail`.

File: cpp/src/io/shp/polygon_shp_reader.cu [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+using namespace std;
+
+namespace cuspatial
+{
+ 
+    /*
+     * Read a LinearRing into x/y/size vectors
+    */
+ 
+    void VertexFromLinearRing(OGRLinearRing *poRing, std::vector<double> &aPointX,
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shp_reader.cu]: Pass inputs either by value or by const reference (if copying is non-trivial).
```suggestion
    void VertexFromLinearRing(OGRLinearRing const& poRing, std::vector<double> &aPointX, 
```

File: cpp/src/io/shp/polygon_shp_reader.cu [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        memcpy((void *)(pm.x),(void *)(x_v.data()),pm.num_vertex*sizeof(double));
+        memcpy((void *)(pm.y),(void *)(y_v.data()),pm.num_vertex*sizeof(double));      
+    }
+ 
+    /*
+    * read polygon data from file in ESRI Shapefile format; data type of vertices is fixed to double (GDF_FLOAT64)
+    * see shp_readers.hpp
+    */
+
+    void read_polygon_shp(const char *filename,
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shp_reader.cu]: This is the only function that should be in `namespace cuspatial`

File: cpp/tests/io/read_shapefile_polygon_test.cu [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    gettimeofday(&t0, nullptr);
+    cuspatial::read_polygon_shp(shape_filename.c_str(),&f_pos,&r_pos,&poly_x,&poly_y);
+    std::cout<<"# of polygons= "<<f_pos.size<<std::endl;
+    std::cout<<"# of rings= "<<r_pos.size<<std::endl;
+    std::cout<<"# of vertices= "<<poly_x.size<<std::endl;
+    
+    gettimeofday(&t1, nullptr);
+    float gpu_pip_time1=cuspatial::calc_time("read shapefile time......",t0,t1);
+}
+#endif
```
Comment (harrism) [inline cpp/tests/io/read_shapefile_polygon_test.cu]:
```suggestion
#endif

```

File: cpp/tests/io/read_shapefile_polygon_test.cu [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+#include <utilities/error_utils.hpp>
+#include <tests/utilities/cudf_test_utils.cuh>
+#include <tests/utilities/cudf_test_fixtures.h>
+#include <cuspatial/shp_readers.hpp>
+#include <utility/utility.hpp> 
+
+struct ReadShapefilePolygonTest : public GdfTest 
+{
+};   
+
```
Comment (harrism) [inline cpp/tests/io/read_shapefile_polygon_test.cu]: Only one test?

Please test corner cases, such as empty files, degenerate files, files that aren't empty but have zero for some attributes, etc.

File: cpp/CMakeLists.txt [hunk 10]
```diff
@@ -252,6 +253,7 @@ include_directories("${CMAKE_BINARY_DIR}/include"
 link_directories("${CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES}" # CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES is an undocumented/unsupported variable containing the link directories for nvcc
                  "${CMAKE_BINARY_DIR}/lib"
                  "${FLATBUFFERS_LIBRARY_DIR}"
+                 "${CONDA_LINK_DIRS}"
```
[the remark this answers is not in the document]
Comment (zhangjianting) [reply cpp/CMakeLists.txt]: GDAL is installed under cudf_dev env. This is to add the path to gdal library.

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/CMakeLists.txt]: Please use FindGDAL to find it there.

File: cpp/tests/io/read_shapefile_polygon_test.cu (hunk 9 above)
Comment (zhangjianting) [reply cpp/tests/io/read_shapefile_polygon_test.cu]: More tests will be added after we figure out where to put data files. Currently, tests module is  disabled completely anyway.

File: cpp/src/io/shp/polygon_shp_reader.cu (hunk 7 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shp_reader.cu]: Got it. Will change.

File: cpp/src/io/shp/polygon_shp_reader.cu (hunk 5 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shp_reader.cu]: Got it. Will change.

File: cpp/src/io/shp/polygon_shp_reader.cu (hunk 4 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shp_reader.cu]: Will change.

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (zhangjianting) [reply cpp/CMakeLists.txt]: Since GDAL (and many others in the future) is installed under a conda env, would it be easier to just add CONDA_LINK_DIRS to link_directories? Some of these third-party libraries may not have FindX CMake module.

**Comment by harrism** [comment]
The standard approach to dependencies in cmake is to "find" them. This allows users to install dependencies in other ways, such as using direct installers or OS package managers, or from source, if they prefer. One may need to install gdal, or cuDF, or RMM, or another dependency, from source, for example, in order to debug a subtle bug in interaction with cuspatial.
________________________________

File: cpp/include/cuspatial/shp_readers.hpp (hunk 3 above)
Comment (zhangjianting) [reply cpp/include/cuspatial/shp_readers.hpp]: this is a typo - thanks for the correction.

File: cpp/CMakeLists.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (zhangjianting) [reply cpp/CMakeLists.txt]: Done. 
Left ${CONDA_LINK_DIRS} there. cuDF also added it to link dirs.

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (zhangjianting) [reply cpp/CMakeLists.txt]: Done.

File: cpp/src/io/shp/polygon_shp_reader.cu (hunk 4 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shp_reader.cu]: Done.

File: cpp/src/io/shp/polygon_shp_reader.cu (hunk 5 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shp_reader.cu]: Done. One function (with a new name polygon_from_shapefile ) is left inside cuspatial namespace as it uses struct polygons<T> which is defined inside cuspatial namespace.

File: cpp/src/io/shp/polygon_shp_reader.cu (hunk 7 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shp_reader.cu]: read_polygon_shp-->read_polygon_shapefile. Also see comment above.

File: cpp/CMakeLists.txt (hunk 10 above)
[the remark this answers is not in the document]
Comment (zhangjianting) [reply cpp/CMakeLists.txt]: FindGDAL used.

File: cpp/src/io/shp/polygon_shp_reader.cu (hunk 5 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shp_reader.cu]: changed.

**Comment by achapkowski** [comment]
Has there been any thought to integrate Fiona instead of just a shapefile reader?  Shapefiles are an old outdated storage format.
see: https://github.com/Toblerity/Fiona

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    * Read a polygon shapefile and fill in a polygons structure
+    * ToDo: read associated relational data into a CUDF Table 
+    *
+    * filename: ESRI shapefile name (wtih .shp extension
+    * pm: structure polygons (fixed to double type) to hold polygon data
+    
+    * Note: only the first layer is read - shapefiles have only one layer in GDALDataset model    
+    */
+
+    void polygon_from_shapefile(const char *filename, struct polygons<double>& pm)
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shapefile_reader.cu]: `struct` specifier on an argument is C, not C++.

```suggestion
    void polygon_from_shapefile(const char *filename, polygons<double>& pm)
```

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+    void read_polygon_shapefile(const char *filename,
+                      gdf_column* ply_fpos, gdf_column* ply_rpos,
+                      gdf_column* ply_x, gdf_column* ply_y)
+    {
+        memset(ply_fpos,0,sizeof(gdf_column));
+        memset(ply_rpos,0,sizeof(gdf_column));
+        memset(ply_x,0,sizeof(gdf_column));
+        memset(ply_y,0,sizeof(gdf_column));
+
+        struct polygons<double> pm;
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shapefile_reader.cu]:
```suggestion
        polygons<double> pm;
```

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+        GDALAllRegister();
+        
+        GDALDatasetH hDS = GDALOpenEx( filename, GDAL_OF_VECTOR, NULL, NULL, NULL );
+        CUDF_EXPECTS(hDS!=NULL,"Failed to open ESRI Shapefile dataset");		    
+        OGRLayerH hLayer = GDALDatasetGetLayer( hDS,0 );
+        CUDF_EXPECTS(hLayer!=NULL,"Failed to open the first layer");
+        int num_f=ReadLayer(hLayer,g_len_v,f_len_v,r_len_v,x_v,y_v);
+        CUDF_EXPECTS(num_f>0,"Shapefile must have at lest one polygon");
+        
+        pm.num_group=g_len_v.size();
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shapefile_reader.cu]: Give `polygons` a constructor so you don't have to manually allocate and set every member.

Comment (harrism) [reply cpp/src/io/shp/polygon_shapefile_reader.cu]: And a destructor.

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+        memset(&pm,0,sizeof(pm));
+        polygon_from_shapefile(filename,pm);
+        if (pm.num_feature <=0) return;
+
+        cudaStream_t stream{0};
+        auto exec_policy = rmm::exec_policy(stream)->on(stream);
+
+        int32_t* temp{nullptr};
+        RMM_TRY( RMM_ALLOC(&temp, pm.num_feature * sizeof(int32_t), stream) );
+        CUDA_TRY( cudaMemcpyAsync(temp, pm.feature_length,
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shapefile_reader.cu]: I don't think this copy is necessary. Just scan into temp...

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+                              pm.num_feature * sizeof(int32_t),
+                              cudaMemcpyHostToDevice, stream) );
+        //prefix-sum: len to pos
+        thrust::inclusive_scan(exec_policy, temp, temp + pm.num_feature, temp);
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shapefile_reader.cu]:
```suggestion
        thrust::inclusive_scan(exec_policy, pm.feature_length, pm.feature_length + pm.num_feature, temp);
```

And delete the preceding cudaMemcpyAsync.

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+        gdf_column_view_augmented(ply_fpos, temp, nullptr, pm.num_feature,
+                              GDF_INT32, 0,
+                              gdf_dtype_extra_info{TIME_UNIT_NONE}, "f_pos");
+
+        RMM_TRY( RMM_ALLOC(&temp, pm.num_ring * sizeof(int32_t), stream) );
+        CUDA_TRY( cudaMemcpyAsync(temp, pm.ring_length,
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shapefile_reader.cu]: Same. Remove this copy and scan pm.ring_length into temp.

File: cpp/tests/CMakeLists.txt [hunk 17]
```diff
@@ -79,7 +79,9 @@ include_directories("${CMAKE_BINARY_DIR}/include"
 
 link_directories("${CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES}" # CMAKE_CUDA_IMPLICIT_LINK_DIRECTORIES is an undocumented/unsupported variable containing the link directories for nvcc
                  "${CMAKE_BINARY_DIR}/lib"
-                 "${GTEST_LIBRARY_DIR}"
+                 "${GDAL_LIBRARIES}"
+                 "${CONDA_LINK_DIRS}"
+ 		 "${GTEST_LIBRARY_DIR}"
```
Comment (harrism) [inline cpp/tests/CMakeLists.txt]: Mixing tabs and spaces.

File: cpp/tests/io/read_shapefile_polygon_test.cu (hunk 9 above)
Comment (harrism) [reply cpp/tests/io/read_shapefile_polygon_test.cu]: All you need is a test that reads a (trivial) invalid file and tests that the error reported is correct.

File: cpp/CMakeLists.txt [hunk 18]
```diff
@@ -194,6 +194,20 @@ if (NVSTRINGS_INCLUDE AND NVTEXT_LIBRARY)
...
+
+# - find gdal -------------------------------------------------------------------------------------
+
+find_package(GDAL REQUIRED)
+
+message(STATUS "GDAL: GDAL_LIBRARIES set to ${GDAL_LIBRARIES}")
+message(STATUS "GDAL: GDAL_INCLUDE_DIRS set to ${GDAL_INCLUDE_DIRS}")
+
+if(GDAL_FOUND)
+    message(STATUS "GDAL found in ${GDAL_INCLUDE_DIRS}")
```
[the remark this answers is not in the document]
Comment (zhangjianting) [reply cpp/CMakeLists.txt]: changed.

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 19]
```diff
@@ -0,0 +1,244 @@
...
+        OGRFeatureH hFeat;
+        while( (hFeat = OGR_L_GetNextFeature( layer )) != NULL )
+        {
+            OGRGeometry *poShape=(OGRGeometry *)OGR_F_GetGeometryRef( hFeat );
+            CUDF_EXPECTS(poShape!=NULL,"Invalid Shape");
+            
+            std::vector<double> aPointX;
+            std::vector<double> aPointY;
+            std::vector<int> aPartSize;
+ 	    PolygonFromGeometry( poShape, aPointX, aPointY, aPartSize );
```
[the remark this answers is not in the document]
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shapefile_reader.cu]: changed.

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+            OGRGeometryCollection *poGC = (OGRGeometryCollection *) poShape;
+            for(int i = 0; i < poGC->getNumGeometries(); i++ )
+            {
+                OGRGeometry *poGeom=poGC->getGeometryRef(i);
+                PolygonFromGeometry(poGeom,aPointX, aPointY, aPartSize );
+              }
+        }
+        else if (eFlatType == wkbPolygon)
+            LinearRingFromPolygon(*((OGRPolygon *) poShape),aPointX, aPointY, aPartSize );
+	else
```
[the remark this answers is not in the document]
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shapefile_reader.cu]: changed.

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 21]
```diff
@@ -166,10 +166,12 @@ namespace cuspatial
...
-        CUDF_EXPECTS(pm.group_length!=nullptr&&pm.feature_length!=nullptr&&pm.ring_length!=nullptr,"expecting p_{g,f,r}_len are non-zeron");		    
-        CUDF_EXPECTS(pm.x!=nullptr&&pm.y!=nullptr,"expecting polygon x/y arrays are not nullptr");
+        CUDF_EXPECTS(pm.group_length !=nullptr, "NULL group_length pointer");
+        CUDF_EXPECTS(pm.feature_length != nullptr, "NULL feature_length pointer");
+        CUDF_EXPECTS(pm.ring_length != nullptr, "NULL ring_length pointer");
+        CUDF_EXPECTS(pm.x != nullptr && pm.y != nullptr, "NULL polygon x/y data pointer");
 	      
-        memcpy((void *)(pm.group_length),(void *)(g_len_v.data()),pm.num_group*sizeof(uint32_t));
+        std::copy_n(pm.group_length, pm.num_group, g_len_v.begin());
         memcpy((void *)(pm.feature_length),(void *)(f_len_v.data()),pm.num_feature*sizeof(uint32_t));
```
Comment (harrism) [inline cpp/src/io/shp/polygon_shapefile_reader.cu]: I meant for you to change *all* the memcopies here. :)

File: cpp/src/io/shp/polygon_shapefile_reader.cu (hunk 13 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shapefile_reader.cu]: It is designed to be a struct. memset does the job, just like gdf_column. 
I want to avoid changes to existing code that has already been used previous versions. A more thorough review/redesign/re-implementation can be scheduled in future.

File: cpp/src/io/shp/polygon_shapefile_reader.cu (hunk 21 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shapefile_reader.cu]: Done.

File: cpp/src/io/shp/polygon_shapefile_reader.cu (hunk 15 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shapefile_reader.cu]: pm.feature_length is a host pointer, not a device pointer. Is Thurst supporting scan (and maybe other ops) host vectors into device vectors directly?

File: cpp/tests/io/read_shapefile_polygon_test.cu (hunk 9 above)
Comment (zhangjianting) [reply cpp/tests/io/read_shapefile_polygon_test.cu]: Again, where should these test files be put? I thought we decided to table the issue when a suitable place can be found. Please see my comments on the slack channel.

File: cpp/src/io/shp/polygon_shapefile_reader.cu (hunk 14 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shapefile_reader.cu]: Tested and found that temp is needed. pm pointers are on the host side. Also see the comments below.

File: cpp/src/io/shp/polygon_shapefile_reader.cu (hunk 16 above)
Comment (zhangjianting) [reply cpp/src/io/shp/polygon_shapefile_reader.cu]: Same as above.

File: cpp/tests/CMakeLists.txt (hunk 17 above)
Comment (zhangjianting) [reply cpp/tests/CMakeLists.txt]: changed.

**Comment by thomcom** [comment]
Hey @achapkowski, thanks for your interest! We're investigating which libraries are best to support/mimic/depend upon. The main challenge with GeoJSON/Fiona support is flattening DataSources into GPU representable arrays that we can then perfect fast parallelized algorithms on. So far, simple Shapefiles are as far as we've gotten. :)

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 22]
```diff
@@ -0,0 +1,247 @@
...
+        cudaStream_t stream{0};
+        auto exec_policy = rmm::exec_policy(stream)->on(stream);
+
+        int32_t* temp{nullptr};
+        RMM_TRY( RMM_ALLOC(&temp, pm.num_feature * sizeof(int32_t), stream) );
+        CUDA_TRY( cudaMemcpyAsync(temp, pm.feature_length,
+                              pm.num_feature * sizeof(int32_t),
+                              cudaMemcpyHostToDevice, stream) );
+        //prefix-sum: len to pos
+        thrust::inclusive_scan(exec_policy, temp, temp + pm.num_feature, temp);
```
Comment (thomcom) [inline cpp/src/io/shp/polygon_shapefile_reader.cu]: Should be updated to `exec_policy->on(stream)` as above.

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+        gdf_column_view_augmented(ply_fpos, temp, nullptr, pm.num_feature,
+                              GDF_INT32, 0,
+                              gdf_dtype_extra_info{TIME_UNIT_NONE}, "f_pos");
+
+        RMM_TRY( RMM_ALLOC(&temp, pm.num_ring * sizeof(int32_t), stream) );
+        CUDA_TRY( cudaMemcpyAsync(temp, pm.ring_length,
+                              pm.num_ring * sizeof(int32_t),
+                              cudaMemcpyHostToDevice, stream) );
+        thrust::inclusive_scan(exec_policy, temp, temp + pm.num_feature, temp);
```
Comment (thomcom) [inline cpp/src/io/shp/polygon_shapefile_reader.cu]: `exec_policy->on(stream)`

File: cpp/tests/io/read_shapefile_polygon_test.cu [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+           EXPECT_NEAR(h_polygon.x[i],h_x[i],1e-9);
+        
+        std::vector<double> h_y(h_polygon.num_vertex);
+        EXPECT_EQ(cudaMemcpy(h_y.data(),poly_y.data,
+                             h_polygon.num_vertex* sizeof(double),
+                             cudaMemcpyDeviceToHost),
+                  cudaSuccess);
+        for(size_t i = 0; i<h_polygon.num_vertex;i++)
+            EXPECT_NEAR(h_polygon.y[i],h_y[i],1e-9);
+        return true;
```
Comment (thomcom) [inline cpp/tests/io/read_shapefile_polygon_test.cu]: Whitespace is all over the place here, clean up?

File: cpp/src/io/shp/polygon_shapefile_reader.cu [hunk 25]
```diff
@@ -0,0 +1,247 @@
...
+        memset(ply_x,0,sizeof(gdf_column));
+        memset(ply_y,0,sizeof(gdf_column));
+
+        polygons<double> pm;
+        memset(&pm,0,sizeof(pm));
+        polygon_from_shapefile(filename,pm);
+        if (pm.num_feature <=0) return;
+
+        cudaStream_t stream{0};
+        auto exec_policy = rmm::exec_policy(stream)->on(stream);
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/src/io/shp/polygon_shapefile_reader.cu]: (Just FYI) technically not a leak, but an invalid reference (likely segfault).

**Comment by thomcom** [comment]
rerun tests

**Comment by thomcom** [comment]
rerun tests

**Comment by harrism** [comment]
The missing file used to be in cudf's src/utilities directory but recently (6 days ago) moved to include/cudf/detail/integer_utils.hpp  For some reason CI doesn't have the latest cuDF?

File: cpp/tests/io/read_shapefile_polygon_test.cu [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+    h_polygon.x=new double[h_polygon.num_vertex]{-10,   5, 5, -10, -10,0, 10, 10,  0, 0};
+    h_polygon.y=new double[h_polygon.num_vertex]{-10, -10, 5,   5, -10,0,  0, 10, 10, 0};
+
+    gdf_column f_pos,r_pos,poly_x,poly_y;
+    cuspatial::read_polygon_shapefile(shape_filename.c_str(),&f_pos,&r_pos,&poly_x,&poly_y);
+
+    CUDF_EXPECTS(this->check_polygon(h_polygon,f_pos,r_pos,poly_x,poly_y),"polygon readout mismatches expected");
+}
+
+TEST_F(ReadShapefilePolygonTest, testITSROI)
```
[the remark this answers is not in the document]
Comment (zhangjianting) [reply cpp/tests/io/read_shapefile_polygon_test.cu]: I will change testITSROI to testReal to avoid confusion.

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/tests/io/read_shapefile_polygon_test.cu]: The meaning of `testReal` isn't really clearer...

[the remark this answers is not in the document]
Comment (zhangjianting) [reply cpp/tests/io/read_shapefile_polygon_test.cu]: It might be better to just remove the test as unit tests should not involve real datasets.

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/tests/io/read_shapefile_polygon_test.cu]: Why shouldn't they?

File: README.md [hunk 27]
```diff
@@ -47,62 +49,37 @@ environment created in step 3 is active.
...
-python  $CUSPATIAL_HOME/python/cuspatial/cuspatial/tests/pip2_test_soa_toy.py
-```
-
-However, many test code uses real data from an ITS (Intelligent Transportation
+Finally, many test/demo code uses real data from an ITS (Intelligent Transportation
 System) application. You will need to follow instructions at
 [data/README.md](./data/README.md) to generate data for these test code.
 Alternatively, you can download the preprocessed data ("locust.*",
 "its_4326_roi.*", "itsroi.ply" and "its_camera_2.csv") from 
 [here](https://nvidia-my.sharepoint.com/:u:/p/jiantingz/EdHR7qlaRSVPtw46XYVR9sQBjCcnUHygCuPUC3Hf8gW73A?e=LCr9nK).
```
[the remark this answers is not in the document]
Comment (zhangjianting) [reply README.md]: I have removed the paragraph. Again, unit tests (both C++ and Python) should not involve real datasets.

File: cpp/tests/io/read_shapefile_polygon_test.cu (hunk 26 above)
[the remark this answers is not in the document]
Comment (zhangjianting) [reply cpp/tests/io/read_shapefile_polygon_test.cu]: I think testing on real datasets should be in demo/benchmarking section. I did not see cudf uses real data in unit tests.

[the remark this answers is not in the document]
Comment (zhangjianting) [reply cpp/tests/io/read_shapefile_polygon_test.cu]: A PR might be needed to move tests that use real datasets out from respective modules. The are currently disabled by using "#if 0", which is far from elegant.

**Comment by harrism** [comment]
I think gtests with some real data would be good to have, assuming (down)loading the data doesn't take a long time. "demos" don't get run by CI, so they aren't very useful for testing.

**Comment by zhangjianting** [comment]
> I think gtests with some real data would be good to have, assuming (down)loading the data doesn't take a long time. "demos" don't get run by CI, so they aren't very useful for testing.

There are four tests already. The last two tests have pretty much the same code except filenames. Adding another test with same test code does not provide additional benefits.

**Comment by zhangjianting** [comment]
rerun tests

**Comment by thomcom** [comment]
rerun tests

**Comment by thomcom** [comment]
rerun tests

**Comment by thomcom** [comment]
rerun tests
